File: hello/hello.js
// Copyright 2011 Google Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// before this script, be sure that the following is set:
//    topDiv -- a jQuery object for the top div
//    storage.get & storage.put -- save and restore state

var languageMap = {
  ar: 'مرحبا العالم!',
  cy: 'Helo byd!',
  el: 'Γεια σας κόσμο!',
  en: 'Hello World!',
  es: '¡Hola, mundo!',
  fi: 'Moikka maailma!',
  is: 'Halló heimur!',
  it: 'Ciao mondo!',
  iw: 'שלום עולם!',
  ja: 'こんにちは、世界！',
  ko: '안녕하세요!',
  hi: 'नमस्ते विश्व!',
  nl: 'Hallo wereld!',
  ru: 'Привет мир!',
  sr: 'Здраво свете!',
  sv: 'Hej världen!',
  sw: 'Hello ulimwengu!',
  th: 'สวัสดีชาวโลก',
  vi: 'Xin chào thế giới!',
  yi: 'העלא וועלט!',
  zh: '世界您好！'
};

var languages = [];
for (var l in languageMap) { languages.push(l); }

var setLang = function(l) {
  topDiv.find('p').text(languageMap[l]);
  if (storage.get() != l) {
     storage.put(l);
  }
}

topDiv.find('a').click(function() {
  var l = languages[Math.floor(Math.random() * languages.length)];
  setLang(l);
  return false;
});

setLang(storage.get() || 'en');

File: hello/hello.py
# Copyright 2011 Google Inc. All Rights Reserved.
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#     http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

#!/usr/bin/env python

import datetime
import logging
import os
import sys
import uuid
from google.appengine.ext import db
from google.appengine.ext import webapp
from google.appengine.ext.webapp.util import run_wsgi_app
from lib.py import belay


os.environ['DJANGO_SETTINGS_MODULE'] = 'settings'
from django.template.loader import render_to_string

def server_url(path):
  return belay.this_server_url_prefix() + path

class HelloData(db.Model):
  lang = db.TextProperty()


class ViewHandler(belay.BcapHandler):
  def get(self):
    content = \
      render_to_string("hello-gadget.html", { 'url': server_url("/hello.css") })
    self.xhr_content(content, "text/html;charset=UTF-8")


class GenerateHandler(belay.BcapHandler):
  def get(self):
    hello = HelloData()
    hello.put()
    self.bcapResponse(belay.regrant(LaunchHandler, hello))


class GenerateInstanceHandler(belay.BcapHandler):
  def get(self):
    hello = HelloData()
    hello.put()
    self.bcapResponse({
      'launch': belay.regrant(LaunchHandler, hello),
      'icon': server_url("/tool-hello.png"), 
      'name': 'Hello'
    })


class LaunchHandler(belay.CapHandler):
  def get(self):
    self.bcapResponse({
        'page': {
          'html': server_url("/hello-belay.html"),
          'window': { 'width': 300, 'height': 250 }
        },
        'gadget': {
          'html': server_url("/view/gadget"),
          'scripts': [ server_url("/hello.js") ]
        },
        'info': { 
          'lang': self.get_entity().lang,
          'setLang': belay.regrant(LaunchHandler, self.get_entity())

        }
      })

  def put(self):
    hello = self.get_entity()
    hello.lang = self.bcapRequest()
    hello.put()
    self.bcapNullResponse()


application = webapp.WSGIApplication(
  [(r'/cap/.*', belay.ProxyHandler),
  ('/belay/generate', GenerateHandler),
  ('/belay/generate-instance', GenerateInstanceHandler),
  ('/view/gadget', ViewHandler),
  ],
  debug=True)
  

belay.set_handlers(
  '/cap',
  [ ('/belay/launch', LaunchHandler),
  ])

def main():
  logging.getLogger().setLevel(logging.DEBUG)
  run_wsgi_app(application)

if __name__ == "__main__":
  main()
